### PR TITLE
nRF54L: Fix LFXO and HFXO INTCAP code calculations

### PR DESCRIPTION
**Pull request by ppryga-nordic**
**soc: nrf54l15: Fix LFXO INTCAP code calculations**

There were two errors in calculations of LFXO INTCAP code:
- The value provided by DTS files is internal desired capacitance.
  The value from DTS has to be "encoded" before use in INTCAP
  calculations formula. The formula for encoding is:
  CAPACITANCE_CODE = (<desired_value> - 4pF) / 0.5
  Subtract of 4 is related with lowest value in the allowed range.
  Division by 0.5 is related with change to steps size.
  In former code the subtration of 4pF was missing.

- The mid_val calcuation was wrong due offset_k left shift by 4.
  It should be left shift by 3 to get total left shift of 9.
  That matches the left shift of former part of the equation.
  Final integer value was calculated by right shift 10, it should
  be right shift 9. Then rounding was done by use of mod by (1 << 10)
  It should be mod by (1 << 9) and compared with (1 << 8), that is
  half of 0-512 range.

**soc: nrf54l15: Fix HFXO INTCAP code calculation rounding error**

The implementation of formula, to calculate HFXO INTCAP code,
had a rounding error. That may lead so small deviation of HFXO
clock accuracy.

The IPS formula uses capacitance values in piko Fartd units.
That requires use of floating point data types. To avoid that
implementation of the formula uses femto Farats (1000 smaller unit).

In the former implementation conversion from femto Farad to piko Farad
was done just after reading of the desired capacitance from DTS.

To make sure the calculations are correct the change of unit must be
done at very end. Also rounding must be applied.

Also the formula was split and more comments were added to make the
implementation clear.

The commit fixes the implementation of the IPS formula.

**drivers: clock_control: Remove XTAL accuracy change for nRF54L**

There was an error in calculation of LFXO INTCAP code that prevented
obtaining desired XTAL accuracy below 50PPM. The error was fixed
the accuracy should be within expected range.

### CODE REVIEW

**Comment by ppryga-nordic** [comment]
Fixed compliance errors.

**Review by martintv — APPROVED** [review]

**Review by nordic-krch — DISMISSED** [review]

**Comment by nashif** [comment]
[nrf fromlist] ?

File: soc/nordic/nrf54l/soc.c [hunk 1]
```diff
@@ -65,16 +65,36 @@ static int nordicsemi_nrf54l_init(void)
...
+	uint32_t mid_val = (cap_val_encoded - 4UL) * (uint32_t)(slope_k + 392UL)
+			   + (offset_k << 3UL);
+
+	/* Get integer part of the INTCAP code */
+	uint32_t lfxo_intcap = mid_val >> 9UL;
+
+	/* Round based on fractional part */
+	uint32_t int_cap_fract = (mid_val % (uint32_t)(1 << 9));
+
+	if (int_cap_fract >= (uint32_t)(1 << 8)) {
```
Comment (anangl) [inline soc/nordic/nrf54l/soc.c:93]: How about:
```suggestion
	if ((mid_val & BIT_MASK(9)) >= (BIT_MASK(9) / 2) {
```

**Review by anangl — DISMISSED** [review]

**Comment by ppryga-nordic** [comment]
> [nrf fromlist] ?

sorry my bad - removed

File: soc/nordic/nrf54l/soc.c (hunk 1 above)
Comment (ppryga-nordic) [reply soc/nordic/nrf54l/soc.c:93]: done

**Review by anangl — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**